enhancements/fail-forward-upgrades.md

```diff
@@ -0,0 +1,252 @@
+
+##### Supporting "Fail Forward" Upgrades when one or more CSVs have failed
+
+Let's review what steps must be taken to recover from a failed InstallPlan today:
+1. `Operator v1` is being upgraded to `Operator v2`.
+2. The InstallPlan for `Operator v2` fails.
+3. `Operator v3` is added to the catalog which skips `Operator v2` and replaces `Operator v1`.
+4. The user deletes the InstallPlan created for `Operator v2`.
+5. A new InstallPlan is generated for `Operator v3` and the upgrade succeeds.
+
+The enhancement proposes that when the user opts into "fail forward" upgrades that OLM allows new InstallPlans to be generated if:
+- The failed installPlan is referenced by a subscription in the namespace.
+- The catalogSource has been updated to include a new new upgrade for the set of arguments.
+
+In practice, the fourth step from the previous workflow would be removed, meaning that the cluster admins would simply need to update the CatalogSource and wait for the cluster to move past the failed install. We do not need to code special handling for partially applied CSVs as OLM will include any applied CSVs in the set of arguments sent to the resolver.
+
+>BEST PRACTICE NOTE: If a bundle is known to fail, it should be skipped in the upgrade graph using the [skips](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/#skips) or [skipRange](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/#skiprange) feature.
+
+##### Supporting "Fail Forward" Upgrades when an InstallPlan failed
+
+Let's review what steps must be taken to recover from a failed CSV today:
+1. `Operator v1` is being upgraded to `Operator v2`.
+2. The CSV for `Operator v2` enters the FAILED phase.
+3. `Operator v3` is added to the catalog which skips `Operator v2` and replaces `Operator v1`.
+3. The resolver cannot upgrade `Operator v2` while including `Operator v1` in the solution set, upgrade is blocked.
+4. User manually deletes the existing CSVs, a new InstallPlan is generated and approved.
+5. `Operator v3` is installed and the upgrade succeeds.
+
+The enhancement proposes that when the user opts into "fail forward" upgrades that OLM:
+- Does not include CSVs in the REPLACING phase in the set of arguments sent to the Resolver.
+- Allows CSVs to move from the FAILED phase to the REPLACING phase.
+
+In practice, these changes would allow OLM to recover from any number of failed CSVs in an upgrade path. cluster admins would only need to update the CatalogSource with new versions the replace the FAILED CSVs.
+#### Preventing an operator from "Failing Forward" when in the FAILED Phase
```

Comment:
The bundle property was brought up earlier today and we are planning to remove this from the proposal. Unfortunately:
- Behind the scenes, this property only influences which CSVs (arguments) OLM sends to the resolver.
- Once OLM has installed a bundle, the property won't appear on the CSV if the catalog is updated to include it, meaning that if you release a CSV and later deem that it should not support "failing forward", customers that have already installed the CSV can still fail forward.

We will ideally add the ability to opt out of fail forward upgrades with the new APIs if applicable. For now, this feature will be behind a feature-gate and left in the hands of the admins.